pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/validator/CredentialsValidator.java

```diff
@@ -0,0 +1,85 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package org.jenkinsci.plugins.pipeline.modeldefinition.validator;
+
+import hudson.Extension;
+import hudson.model.Queue;
+import hudson.model.Run;
+import org.jenkinsci.plugins.credentialsbinding.impl.CredentialNotFoundException;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTEnvironment;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTEnvironmentValue;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTInternalFunctionCall;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTKey;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTValue;
+import org.jenkinsci.plugins.pipeline.modeldefinition.model.CredentialsBindingHandler;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.Map;
+
+@Extension
+public class CredentialsValidator extends DeclarativeValidatorContributor {
+
+    @CheckForNull
+    @Override
+    public String validateElement(@Nonnull ModelASTEnvironment environment, @CheckForNull FlowExecution execution) {
+        final FlowExecutionOwner owner = execution != null ? execution.getOwner() : null;
+        if (owner != null) {
+            try {
+                final Queue.Executable executable = owner.getExecutable();
+                if (executable != null && executable instanceof Run) {
+                    //This might show the error not exactly in the right place
+                    for (Map.Entry<ModelASTKey, ModelASTEnvironmentValue> entry : environment.getVariables().entrySet()) {
+                        if (entry.getValue() instanceof ModelASTInternalFunctionCall) {
+                            ModelASTInternalFunctionCall func = (ModelASTInternalFunctionCall) entry.getValue();
+                            if (func.getName().equals("credentials")) {
+                                if (func.getArgs().isEmpty()) {
+                                    return Messages.CredentialsValidator_noArgs();
+                                } else {
+                                    final ModelASTValue idValue = func.getArgs().get(0);
+                                    if (idValue.isLiteral() && !idValue.getValue().toString().startsWith("${")) {
+                                        //In this context I guess we can only evaluate literals, not runtime computed values
```

Comment:
Correct. I still need to decide what to do about non-literal values in other places we do validation, like `tools` - right now, we don't allow non-literal values there because it'd break validation, but we've already crossed that bridge with `credentials`, so I'd say that yeah, we can't exactly do validation of non-literals and that's just the price of doing business.